lib/services/suppressed-violations-manager.js

```diff
@@ -0,0 +1,275 @@
+         */
+        const filtered = structuredClone(results);
+        const unused = structuredClone(suppressions);
+
+        for (const result of filtered) {
+            const relativeFilePath = this.getRelativeFilePath(result.filePath);
+
+            if (!suppressions[relativeFilePath]) {
+                continue;
+            }
+
+            const violationsByRule = SuppressedViolationsManager.countViolationsByRule(result.messages);
+
+            for (const ruleId in violationsByRule) {
+                if (!unused[relativeFilePath][ruleId]) {
+                    continue;
+                }
+
+                const suppressionsCount = unused[relativeFilePath][ruleId].count;
+                const violationsCount = violationsByRule[ruleId].count;
+
+                // Suppress messages if the number of violations is less or equal to the suppressions count
+                if (violationsCount <= suppressionsCount) {
+                    SuppressedViolationsManager.suppressMessagesByRule(result, ruleId);
+                }
+
+                // Update the count to match the new number of violations, otherwise remove the rule entirely
+                if (violationsCount < suppressionsCount) {
+                    unused[relativeFilePath][ruleId].count = suppressionsCount - violationsCount;
+                } else {
+                    delete unused[relativeFilePath][ruleId];
+                }
+            }
+
+            // Cleanup files with no rules
+            if (Object.keys(unused[relativeFilePath]).length === 0) {
+                delete unused[relativeFilePath];
+            }
+        }
+
+        return {
+            results: filtered,
+            unused
+        };
+    }
+
+    /**
+     * Loads the suppressions file.
+     * @throws {Error} If the suppressions file cannot be parsed.
+     * @returns {SuppressedViolations} The suppressions.
+     */
+    load() {
+        if (fs.existsSync(this.suppressionsLocation)) {
+            try {
+                return JSON.parse(fs.readFileSync(this.suppressionsLocation, "utf8"));
+            } catch {
+                throw new Error(`Failed to parse suppressions file at ${this.suppressionsLocation}`);
+            }
+        }
+
+        return {};
+    }
+
+    /**
+     * Updates the suppressions file.
+     * @param {SuppressedViolations} suppressions The suppressions to save.
+     * @returns {void}
+     * @private
+     */
+    save(suppressions) {
+        fs.writeFileSync(
+            this.suppressionsLocation,
+            JSON.stringify(suppressions, null, 2)
+        );
+    }
+
+    /**
+     * Counts the violations by rule, ignoring warnings.
+     * @param {LintMessage[]} messages The messages to count.
+     * @returns {Record<string, number>} The number of violations by rule.
+     */
+    static countViolationsByRule(messages) {
+        return messages.reduce((acc, message) => {
```

Comment:
Can we use something more descriptive than `acc` here? Maybe `totals`? `counts`?